selfdrive/ui/qt/setup/setup.cc

```diff
@@ -309,19 +368,42 @@ Setup::Setup(QWidget *parent) : QStackedWidget(parent) {
   downloading_widget = downloading();
   addWidget(downloading_widget);
 
-  failed_widget = download_failed();
-  addWidget(failed_widget);
-
-  QObject::connect(this, &Setup::finished, [=](bool success) {
-    // hide setup on success
-    qDebug() << "finished" << success;
-    if (success) {
-      QTimer::singleShot(3000, this, &QWidget::hide);
-    } else {
-      setCurrentWidget(failed_widget);
+  error_widget = download_error();
+  addWidget(error_widget);
+
+  QLabel *url_label = new QLabel();
+  invalid_url_widget = download_invalid_url(url_label);
+  addWidget(invalid_url_widget);
+
+  QObject::connect(this, &Setup::finished, [=](const DownloadResult &result, const QString &url) {
+    qDebug() << "finished" << result << url;
+    switch (result) {
+      case DownloadResult::ok:
+        // hide setup on success
+        QTimer::singleShot(3000, this, &QWidget::hide);
+        break;
+      case DownloadResult::notExecutable:
+        url_label->setText(url);
+        setCurrentWidget(invalid_url_widget);
+        break;
+      case DownloadResult::error:
+      default:
+        setCurrentWidget(error_widget);
+        break;
     }
   });
 
+  // load fonts
+  QFontDatabase::addApplicationFont("../assets/fonts/Inter-Black.ttf");
```

Comment:
this won't work in setup. are you sure this is necessary on C3? might just be your workstation

Reply:
I did get it to work on C3 and checked the QFontDatabase only listed `JetBrains Mono` after importing it this way, but you are right we should add it to agnos